### PR TITLE
Feature prevent invalid move

### PR DESCRIPTION
**Pull request by paulSanchezMarmelab**
[feature_prevent_invalid_move.webm](https://user-images.githubusercontent.com/130989826/233049590-45188778-68b0-42de-8631-6263b2c4ce41.webm)

https://trello.com/c/lMKMgg6w/7-05-as-hayao-i-want-to-prevent-an-invalid-first-move

- [X] added a way to check if move is allowed or not
- [X] implemented render for allowed tiles
- [X] made some tests

### CODE REVIEW

File: src/move/move.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  [true, false, false, false, false, true],
+  [true, false, false, false, true],
+  [false, true, true, false],
+];
+
+export interface Action {
+  x: number;
+  y: number;
+  isFirstMove: boolean;
+}
```
Comment (slax57) [inline src/move/move.ts:15]: why not call this type `Move` to be more consistent with other variable and fn names?

Also, you will probably need to manipulate only coordinates at some point, so you might as well create a Coordinates type:

```ts
export interface Coordinates {
  x: number;
  y: number;
}

export interface Action {
  coordinates: Coordinates;
  isFirstMove: boolean;
}
```

File: src/move/move.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+export const checkIfMoveIsAllowed = (action: Action): boolean => {
```
Comment (slax57) [inline src/move/move.ts:17]: You should find another name for this method. To me, `checkIfMoveIsAllowed` would imply that this fn is capable of checking any move, while it is actually only responsible of checking the first move.

Unless you planned to add other checks in this method later?

File: src/tile/tile.ts [hunk 3]
```diff
@@ -14,21 +14,23 @@ export type Color = (typeof colors)[number];
 export type Symbol = (typeof symbols)[number];
 
 export type Tile = PlayableTile | NeutralTile;
-
+type Styles = "unallowed" | "allowed";
```
Comment (slax57) [inline src/tile/tile.ts:17]: The name 'Styles' implies a general styling capability, where you actually only support allowed or unallowed. Why not simply name the field `allowedAsFirstMove` (or `allowedAsNextMove` if you plan to reuse this logic for the 2+ turns as well) and set the type to a simple `boolean`?

**Review by slax57 — CHANGES_REQUESTED** [review]

File: src/move/move.ts (hunk 2 above)
Comment (paulSanchezMarmelab) [reply src/move/move.ts:17]: That what I intended to do, given we have [this trello card](https://trello.com/c/7RIfw0X5/3-05-as-hayao-i-want-to-know-if-my-move-respect-constraints-set-up-by-the-previous-move) next in the pipeline.

File: src/gameLoader.ts [hunk 4]
```diff
@@ -25,14 +26,14 @@ const parseBoardFromPath = (filePath: string): Board => {
     fileContent = fs.readFileSync(filePath, "utf8");
   } catch (error) {
     console.error(chalk.red(CANNOT_READ_FILE_ERROR));
-    return null;
+    return initRandomGame();
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/gameLoader.ts]: I agree, user would think that their file successfully loaded and we don't want that. Reverted null and changed errors messages.

File: src/board/board.ts [hunk 5]
```diff
@@ -1,26 +1,45 @@
 import chalk from "chalk";
 import { renderTile, Tile } from "../tile/tile";
+import { checkIfMoveIsAllowed } from "../move/move";
 
 export type NullableTile = Tile | undefined;
 export type Board = NullableTile[][];
 
 const BLANK_CHAR = chalk.black(" ");
+let isFirstRender = false;
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/board/board.ts:9]: Done, now it's tied to turnNumber in gameState

File: src/board/board.ts [hunk 6]
```diff
@@ -1,26 +1,45 @@
...
   return lines.reduce((accumulator, tile) => {
     if (tile == undefined) {
       return accumulator + BLANK_CHAR;
     }
 
+    const isMoveAllowed = checkIfMoveIsAllowed({
+      x,
+      y: lineIndex,
+      isFirstMove: isFirstRender,
+    });
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: Done, you will find it inside computeGameState

File: src/board/board.ts [hunk 7]
```diff
@@ -1,26 +1,45 @@
...
   }, line);
 };
 
 export const renderBoard = (data: Board) => {
-  console.log(chalk.white.bold("-------------"));
-  data.forEach((lines) => {
-    console.log(renderLine(lines));
+  if (isFirstRender === false) {
+    isFirstRender = true;
+  }
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: Same.

File: src/move/move.spec.ts [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+describe("checkIfMoveIsAllowed", () => {
+  it("should block move on unallowed first move tile", () => {
+    expect(checkIfMoveIsAllowed({ x: 1, y: 3, isFirstMove: true })).toBeFalsy();
+  });
+
+  it("should allow move on allowed first move tile", () => {
+    expect(
+      checkIfMoveIsAllowed({ x: 1, y: 3, isFirstMove: false })
+    ).toBeTruthy();
+  });
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/move/move.spec.ts]: Done

File: src/move/move.ts (hunk 1 above)
Comment (paulSanchezMarmelab) [reply src/move/move.ts]: I just kept coordinates and moved whole "actions" concept to gameState

File: src/tile/tile.ts (hunk 3 above)
Comment (paulSanchezMarmelab) [reply src/tile/tile.ts]: True ! I replaced it with "moveAllowed" property.

File: src/board/board.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (slax57) [reply src/board/board.ts:9]: Ok, seems better indeed. So... you can probably remove this line now right?

File: src/board/board.ts [hunk 9]
```diff
@@ -1,26 +1,31 @@
 import chalk from "chalk";
 import { renderTile, Tile } from "../tile/tile";
+import { checkIfMoveIsAllowed } from "../move/move";
```
Comment (slax57) [inline src/board/board.ts:3]: unused import

File: src/board/board.ts [hunk 10]
```diff
@@ -1,26 +1,31 @@
...
 export type NullableTile = Tile | undefined;
 export type Board = NullableTile[][];
 
 const BLANK_CHAR = chalk.black(" ");
+let isFirstRender = false;
 
-export const renderLine = (lines: Tile[]): string => {
+export const renderLine = (
+  lines: NullableTile[],
+  lineIndex: number
```
Comment (slax57) [inline src/board/board.ts:13]: `lineIndex` is unused

File: src/board/board.ts [hunk 11]
```diff
@@ -1,26 +1,31 @@
...
+): string => {
   const line = "";
+  let x = 0;
```
Comment (slax57) [inline src/board/board.ts:16]: I don't believe `x` is used

File: src/board/board.ts [hunk 12]
```diff
@@ -1,26 +1,31 @@
...
+    x++;
     return accumulator + renderTile(tile);
   }, line);
 };
 
 export const renderBoard = (data: Board) => {
-  console.log(chalk.white.bold("-------------"));
-  data.forEach((lines) => {
-    console.log(renderLine(lines));
+  data.forEach((lines, y) => {
```
Comment (slax57) [inline src/board/board.ts:28]: You removed the `console.log(chalk.white.bold("-------------"));` instructions. Is this intentional? :thinking:

File: src/board/board.ts [hunk 13]
```diff
@@ -1,26 +1,31 @@
...
+    console.log(renderLine(lines, y));
```
Comment (slax57) [inline src/board/board.ts:29]: `renderLine` does no longer need the second parameter

File: src/gameState.ts [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+  currentPlayer: Player;
+  winner?: Player;
+  selectedMoveCordinates?: Coordinates;
+}
+export type GameActionName =
+  | "SELECT_TILE"
+  | "FORFEIT"
+  | "WAITING"
+  | "END_TURN"
+  | "LOAD_BOARD";
```
Comment (slax57) [inline src/gameState.ts:24]: All these do not look like actions to me. E.g. `SELECT_TILE` could be an action but `WAITING` seems like a state (or status) to me.

Also, you are anticipating on states that are out of scope of this PR (like `FORFEIT`). No need to implement them now. Remember that the client can change goals and priorities at any time, so just focus on the current feature you are implementing.

File: src/gameState.ts [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+  selectedMoveCoordinates?: Coordinates;
+  newBoard?: Board;
+}
+
+const setAllowedTilesOnBoard = (gameState: GameState): Board => {
+  if (gameState.board == null) {
+    return null;
+  }
+
+  if (gameState.turnNumber === 1) {
```
Comment (slax57) [inline src/gameState.ts:36]: You are checking that `gameState.turnNumber === 1` twice, here and at line 48. Is this intended?

File: src/gameState.ts [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+          gameState.turnNumber === 1
+        );
+      })
+    );
+    return newBoard;
+  }
+};
+
+export const computeGameState = (
+  gameState: GameState | undefined,
```
Comment (slax57) [inline src/gameState.ts:57]: Why do you allow `gameState` to be undefined?

Also, you never check that  `gameState` is actually defined later in the code.

File: src/gameState.ts [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+      gameState.turnNumber === 1
+    );
+
+    if (isMoveAllowed) {
+      return {
+        ...gameState,
+        selectedMoveCordinates: action.selectedMoveCoordinates,
+      };
+    }
+  }
```
Comment (slax57) [inline src/gameState.ts:72]: This is partial duplicate of the code implemented in #7 . You can probably clean most of it after rebasing onto `main`.

File: src/gameState.ts [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+  if (action.name === "LOAD_BOARD") {
+    const board: Board = loadGameConfigFromFile();
+    const newGameState = { ...initialGameState, board: board };
+
+    return { ...newGameState, board: setAllowedTilesOnBoard(newGameState) };
+  }
```
Comment (slax57) [inline src/gameState.ts:79]: I'm a bit surprised to see board initialisation logic here. IMO `computeGameState` should be used to update the game state from a previously (already initialized) game state.

Creating the initial board can be done at program startup, in `index.ts` for example. This is the approach taken by #7  if I'm not mistaken so we should keep it :slightly_smiling_face:

**Review by slax57 — CHANGES_REQUESTED** [review]

File: src/gameState.ts (hunk 18 above)
Comment (paulSanchezMarmelab) [reply src/gameState.ts]: Same as https://github.com/marmelab/kamon/pull/7#discussion_r1172496904

File: src/gameState.ts (hunk 16 above)
Comment (paulSanchezMarmelab) [reply src/gameState.ts]: No longer exists since merge with #8

File: src/gameState.ts [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+  newBoard?: Board;
+}
+
+const setAllowedTilesOnBoard = (gameState: GameState): Board => {
+  if (gameState.board == null) {
+    return null;
+  }
+
+  if (gameState.turnNumber === 1) {
+    let newBoard: Board = JSON.parse(JSON.stringify(gameState.board));
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/gameState.ts]: No longer exists since merge with https://github.com/marmelab/kamon/pull/8

File: src/gameState.ts (hunk 15 above)
Comment (paulSanchezMarmelab) [reply src/gameState.ts]: No longer exists since merge with https://github.com/marmelab/kamon/pull/8

File: src/board/board.ts [hunk 20]
```diff
@@ -51,6 +97,8 @@ export const updateBoardState = (
   );
 
   board[lineIndex][tileIndex] = tile;
+  console.log(gameState.turnNumber);
```
Comment (slax57) [inline src/board/board.ts:100]: `console.log` spotted

File: src/gameState.ts (hunk 14 above)
Comment (paulSanchezMarmelab) [reply src/gameState.ts]: No longer exists since merge with https://github.com/marmelab/kamon/pull/8

File: src/board/board.ts (hunk 13 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts:39]: I removed it when merging with #8

File: src/board/board.ts (hunk 12 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts:38]: No, it should be okay now !

File: src/board/board.ts (hunk 11 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts:26]: I removed it.

**Review by slax57 — CHANGES_REQUESTED** [review]

File: src/board/board.ts (hunk 10 above)
Comment (slax57) [reply src/board/board.ts:23]: I still see it :shrug:

File: src/board/board.ts (hunk 9 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: I removed it and will see later to make eslint warn me of any unused import.

File: src/board/board.ts (hunk 11 above)
Comment (slax57) [reply src/board/board.ts:26]: I still see it

File: src/board/board.ts (hunk 12 above)
Comment (slax57) [reply src/board/board.ts:38]: You kept only one out of the two, is this intentional?

File: src/board/board.ts (hunk 13 above)
Comment (slax57) [reply src/board/board.ts:39]: I still see it

File: src/gameState.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (slax57) [reply src/gameState.ts]: I still see it in `board.ts`

File: src/board/board.ts [hunk 21]
```diff
@@ -10,32 +10,72 @@ import {
...
-  data.forEach((lines) => {
-    console.log(renderLine(lines));
+  data.forEach((lines, y) => {
+    console.log(renderLine(lines, y));
   });
   console.log(chalk.white.bold("-------------"));
 };
 
+const clearAllowedTilesHighlight = (board: Board): Board => {
+  let newBoard = JSON.parse(JSON.stringify(board));
```
Comment (slax57) [inline src/board/board.ts:45]:
```suggestion
  const newBoard = JSON.parse(JSON.stringify(board));
```

**Review by slax57 — CHANGES_REQUESTED** [review]

**Comment by slax57** [comment]
You should remove the file `move.spec.ts` if it's empty, because otherwise Jest fails with `Your test suite must contain at least one test.`

File: src/gameState.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/gameState.ts]: Indeed, there were more. I fixed it.

File: src/index.ts [hunk 22]
```diff
@@ -1,8 +1,43 @@
...
+const highlightedInitialBoard = highlightAllowedTiles(gameConfig, gameState);
+renderTurnDisplay(gameState.turnNumber);
+renderBoard(highlightedInitialBoard);
+
+(async () => {
+  while (gameState.isRunning) {
+    const action = await prompt(gameState, highlightedInitialBoard);
+    const userCanMove = checkUserMove(gameConfig, action, gameState);
+    gameState = userCanMove.gameState;
+    if (!userCanMove.allowedMove) continue;
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/index.ts]: Done. I also changed global variable gameState's name to avoid conflict.

File: src/tile/tile.ts [hunk 23]
```diff
@@ -1,5 +1,6 @@
 import chalk from "chalk";
-import { Player } from "../player";
+import { Player } from "../player/player";
+import { Board } from "../board/board";
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/tile/tile.ts]: Player seems to be part of PlayableTile on my side and Board is used all across this file. Did I miss something ?

File: src/move/move.ts [hunk 24]
```diff
@@ -1,27 +1,105 @@
...
+
+export const ALLOWED_FIRST_MOVES_ONLY_TILES = [
+  [false, true, true, false],
+  [true, false, false, false, true],
+  [true, false, false, false, false, true],
+  [false, false, false, false, false, false, false],
+  [true, false, false, false, false, true],
+  [true, false, false, false, true],
+  [false, true, true, false],
 ];
```
[the remark this answers is not in the document]
Comment (paulSanchezMarmelab) [reply src/move/move.ts]: ALLOWED_FIRST_MOVES_WITH_EMPTY_SPACES is used in rendering function (in board.ts). They take account of spacing characters and have to render them.

ALLOWED_FIRST_MOVES_ONLY_TILES is for game logic and user input checking (in move.ts). These only take account of non-null tiles.

I did this as a simple and fast solution. To improve readability I could probably move ALLOWED_FIRST_MOVES_WITH_EMPTY_SPACES to board.ts.

File: src/board/board.ts (hunk 21 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: Done.

File: src/board/board.ts (hunk 20 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: I deleted it.

File: src/board/board.ts (hunk 13 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: This time it should be gone!

File: src/board/board.ts (hunk 11 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: Same, it should be deleted, along with all usage of x in the function.

File: src/index.ts [hunk 25]
```diff
@@ -1,32 +1,51 @@
...
-    if (!userCanMove.allowedMove) continue;
+  while (currentGameState.isRunning) {
+    const action = await prompt(currentGameState, highlightedInitialBoard);
+    const userCanMove = checkUserMove(gameConfig, action, currentGameState);
+
+    const { gameState, allowedMove } = checkUserMove(
+      gameConfig,
+      action,
+      currentGameState
+    );
```
Comment (slax57) [inline src/index.ts:35]: Duplicate (and probably unnecessary) call to `checkUserMove`

File: src/move/move.ts (hunk 24 above)
[the remark this answers is not in the document]
Comment (slax57) [reply src/move/move.ts]: Okay let's not worry about that for now, but indeed this could be improved.

File: src/tile/tile.ts (hunk 23 above)
[the remark this answers is not in the document]
Comment (slax57) [reply src/tile/tile.ts]: my bad, `Board` is indeed used

File: src/board/board.ts (hunk 12 above)
Comment (paulSanchezMarmelab) [reply src/board/board.ts]: Nope, it's back !

File: src/index.ts (hunk 25 above)
Comment (paulSanchezMarmelab) [reply src/index.ts]: It's gone.

File: src/board/board.spec.ts [hunk 26]
```diff
@@ -5,7 +5,7 @@ describe("renderLine", () => {
   it("should render a line with blank characters", () => {
     const tile: Tile = { color: "blue", symbol: "A" };
     const line: NullableTile[] = [undefined, tile, undefined];
-    expect(renderLine(line)).toMatch(/\s.*A.*\s/);
+    expect(renderLine(line, 0)).toMatch(/\s.*A.*\s/);
```
Comment (slax57) [inline src/board/board.spec.ts:8]:
```suggestion
    expect(renderLine(line)).toMatch(/\s.*A.*\s/);
```

**Review by slax57 — CHANGES_REQUESTED** [review]

File: src/board/board.spec.ts (hunk 26 above)
Comment (paulSanchezMarmelab) [reply src/board/board.spec.ts]: Fixed.

**Review by slax57 — APPROVED** [review]
